profiling/memory-usage/kp_memory_usage.cpp

```diff
@@ -128,6 +142,34 @@ void kokkosp_deallocate_data(const SpaceHandle space, const char* /*label*/,
   }
 }
 
+void kokkosp_begin_deep_copy(SpaceHandle dst_handle, const char* /* dst_name */,
+                             const void* /* dst_ptr */, SpaceHandle src_handle,
+                             const char* /* src_name */,
+                             const void* /* src_ptr */, uint64_t size) {
+  std::lock_guard<std::mutex> lock(m);
+
+  int space_dst = num_spaces;
+  for (int s = 0; s < num_spaces; s++)
+    if (strcmp(space_name[s], dst_handle.name) == 0) space_dst = s;
+
+  if (space_dst == num_spaces) {
+    strncpy(space_name[num_spaces], dst_handle.name, 64);
+    num_spaces++;
+  }
+
+  int space_src = num_spaces;
+  for (int s = 0; s < num_spaces; s++)
+    if (strcmp(space_name[s], src_handle.name) == 0) space_src = s;
+
+  if (space_src == num_spaces) {
+    strncpy(space_name[num_spaces], src_handle.name, 64);
+    num_spaces++;
+  }
+  totalMemoryTransferred[space_dst][space_src] += size;
+}
+
+void kokkosp_end_deep_copy() { std::lock_guard<std::mutex> lock(m); }
```

Comment:
Why do we need to acquire the mutex here?

Reply:
Thanks for pointing it out. This `lock_guard` is not needed and it could degrade performance for this tool.

When I was first outlining this PR, I was thinking I would need to add a code block for post-processing at the end of a `Kokkos::deep_copy()` that needs a lock around it.

I will take this out.